### PR TITLE
Add Recursive search

### PR DESCRIPTION
**Pull request by vives**

### CODE REVIEW

File: fileconnector/fileconnector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/FileSearch.java [hunk 1]
```diff
@@ -100,4 +120,59 @@ private void search(String source, String filePattern, MessageContext messageCon
...
+                    }
+                } else if (file.getType() == FileType.FOLDER) {
+                    searchSubFolders(file, filePattern, messageContext, factory, result, ns);
+                }
+            }
+        } catch (IOException e) {
+            handleException("Unable to search a file in sub folder.", e, messageContext);
+        }
+    }
 }
```
Comment (malakasilva) [inline fileconnector/fileconnector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/FileSearch.java]: Don't you have to close file objects after using? Not sure please check this. Better do a load test for this feature.

File: fileconnector/fileconnector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/FileSearch.java [hunk 2]
```diff
@@ -76,15 +82,29 @@ private void search(String source, String filePattern, MessageContext messageCon
...
+                                    getBaseName().toLowerCase())) {
+                                outputResult = child.getName().getBaseName();
+                                OMElement messageElement = factory.createOMElement(FileConstants.FILE,
+                                        ns);
+                                messageElement.setText(outputResult);
+                                result.addChild(messageElement);
+                            } else if (child.getType() == FileType.FOLDER) {
+                                searchSubFolders(child, filePattern, messageContext, factory, result, ns);
+                            }
                         }
```
Comment (malakasilva) [inline fileconnector/fileconnector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/FileSearch.java]: Avoid duplication of code. Following can be used to implement the above logic.

                      if (child.getType() == FileType.FILE && fpm.validate(child.getName().
                                     getBaseName().toLowerCase())) {
                                 outputResult = child.getName().getPath();
                                 OMElement messageElement = factory.createOMElement(FileConstants.FILE,
                                         ns);
                                 messageElement.setText(outputResult);
                                 result.addChild(messageElement);
                      } else if ( child.getType() == FileType.FOLDER && "false".equals(recursiveSearch)){
                                 searchSubFolders(child, filePattern, messageContext, factory, result, ns);
                      }

File: fileconnector/fileconnector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/FileSearch.java [hunk 3]
```diff
@@ -100,4 +123,75 @@ private void search(String source, String filePattern, MessageContext messageCon
             }
         }
     }
+
+    /**
+     *
```
Comment (pamod) [inline fileconnector/fileconnector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/FileSearch.java]: Shall we explain what this class does ?

Comment (vives) [reply fileconnector/fileconnector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/FileSearch.java]: Fixed in https://github.com/wso2-extensions/esb-connector-file/pull/28

**Pull request merged**